### PR TITLE
add validation on complete order

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| New feature?     | yes
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 1]
```diff
@@ -954,6 +960,21 @@ public function iShouldNotBeAbleToSpecifyProvinceNameManuallyForBillingAddress()
         $this->assertProvinceMessage('billingAddress');
     }
 
+    /**
+     * @Then /^I should be notified that (this product) does not have sufficient stock$/
+     * @Then I should be notified that :product does not have sufficient stock
+     */
+    public function iShouldBeNotifiedThatThisProductDoesNotHaveSufficientStock(ProductInterface $product)
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:967]: Missing `: void` 💃

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        /** @var OrderProductEligibility $constraint */
+        Assert::isInstanceOf($constraint, OrderProductInStockEligibility::class);
+
+        $order = $this->orderRepository->findOneByTokenValue($value->getOrderTokenValue());
+
+        /** @var OrderInterface $order */
+        Assert::isInstanceOf($order, OrderInterface::class);
+
+        foreach ($order->getItems() as $orderItem) {
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php:57]: 😱

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        /** @var OrderProductEligibility $constraint */
+        Assert::isInstanceOf($constraint, OrderProductInStockEligibility::class);
+
+        $order = $this->orderRepository->findOneByTokenValue($value->getOrderTokenValue());
+
+        /** @var OrderInterface $order */
+        Assert::isInstanceOf($order, OrderInterface::class);
+
+        foreach ($order->getItems() as $orderItem) {
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php:56]: We should probably typehint `$orderItem` with `OrderItemInterface` from Core

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\Component\Inventory\Checker\AvailabilityCheckerInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidatorInterface;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+
+final class OrderProductInStockEligibilityValidatorSpec extends ObjectBehavior
+{
+    function let(OrderRepositoryInterface $orderRepository, AvailabilityCheckerInterface $availabilityChecker) {
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php:31]: coding-standards 💃

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        $executionContext
+            ->addViolation('sylius.product_variant.product_variant_with_name_not_sufficient', ['%productVariantName%' => 'variant name'])
+            ->shouldBeCalled()
+        ;
+
+        $this->validate($command, new OrderProductInStockEligibility());
+    }
+
+    function it_does_nothing_if_product_variant_have_sufficient_stock(
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php:85]: `have` -> `has`

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        $command = new CompleteOrder();
+        $command->setOrderTokenValue('cartToken');
+
+        $orderRepository->findOneByTokenValue('cartToken')->willReturn($order);
+
+        $order->getItems()->willReturn($orderItems->getWrappedObject());
+
+        $orderItems->getIterator()->willReturn(new \ArrayIterator([$orderItem->getWrappedObject()]));
+
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php:105]: Doubled blank line

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        $this->initialize($executionContext);
+
+        $command = new CompleteOrder();
+        $command->setOrderTokenValue('cartToken');
+
+        $orderRepository->findOneByTokenValue('cartToken')->willReturn($order);
+
+        $order->getItems()->willReturn($orderItems->getWrappedObject());
+
+        $orderItems->getIterator()->willReturn(new \ArrayIterator([$orderItem->getWrappedObject()]));
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php:103]: Why not `ArrayCollection`?

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibility.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+final class OrderProductInStockEligibility extends Constraint
+{
+    public $message = 'sylius.product_variant.product_variant_with_name_not_sufficient';
+
+    public function validatedBy()
+    {
+        return 'sylius_api_validator_order_product_in_stock_eligibility';
+    }
+
+    public function getTargets()
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibility.php:28]: missing return types 🚀

File: src/Sylius/Behat/Resources/config/suites/api/inventory/checkout_inventory.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+                - sylius.behat.context.transform.lexical
+                - sylius.behat.context.transform.order
+                - sylius.behat.context.transform.payment
+                - sylius.behat.context.transform.product
+                - sylius.behat.context.transform.shared_storage
+                - sylius.behat.context.transform.shipping_method
+                - sylius.behat.context.transform.cart
+                    
+                - sylius.behat.context.api.shop.checkout
+                - sylius.behat.context.api.shop.cart
```
Comment (arti0090) [inline src/Sylius/Behat/Resources/config/suites/api/inventory/checkout_inventory.yaml:27]:
```suggestion
                - sylius.behat.context.api.shop.cart
                - sylius.behat.context.api.shop.checkout
```

File: src/Sylius/Behat/Resources/config/suites/api/inventory/checkout_inventory.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+                - sylius.behat.context.transform.address
+                - sylius.behat.context.transform.customer
+                - sylius.behat.context.transform.lexical
+                - sylius.behat.context.transform.order
+                - sylius.behat.context.transform.payment
+                - sylius.behat.context.transform.product
+                - sylius.behat.context.transform.shared_storage
+                - sylius.behat.context.transform.shipping_method
+                - sylius.behat.context.transform.cart
```
Comment (arti0090) [inline src/Sylius/Behat/Resources/config/suites/api/inventory/checkout_inventory.yaml:24]: `sylius.behat.context.transform.cart` before `sylius.behat.context.transform.customer`

File: src/Sylius/Behat/Resources/config/suites/api/inventory/checkout_inventory.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    suites:
+        api_checkout_inventory:
+            contexts:
+                - sylius.behat.context.hook.doctrine_orm
+                    
+                - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.shipping
+                - sylius.behat.context.setup.payment
+                - sylius.behat.context.setup.shop_api_security
```
Comment (arti0090) [inline src/Sylius/Behat/Resources/config/suites/api/inventory/checkout_inventory.yaml:14]: alphabetical order is missing 😢

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibility.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/** @experimental */
+final class OrderProductInStockEligibility extends Constraint
+{
+    public $message = 'sylius.product_variant.product_variant_with_name_not_sufficient';
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibility.php:21]:
```suggestion
    /** @var string */
    public $message = 'sylius.product_variant.product_variant_with_name_not_sufficient';
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        AvailabilityCheckerInterface $availabilityChecker
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->availabilityChecker = $availabilityChecker;
+    }
+
+    /**
+     * @throws \InvalidArgumentException
+     */
+    public function validate($value, Constraint $constraint)
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php:44]:
```suggestion
    public function validate($value, Constraint $constraint)
```

I don't remember we were using in any other of validators this comment.

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        $this->orderRepository = $orderRepository;
+        $this->availabilityChecker = $availabilityChecker;
+    }
+
+    public function validate($value, Constraint $constraint)
+    {
+        Assert::isInstanceOf($value, OrderTokenValueAwareInterface::class);
+
+        /** @var OrderProductEligibility $constraint */
+        Assert::isInstanceOf($constraint, OrderProductInStockEligibility::class);
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibilityValidator.php:47]:
```suggestion
        /** @var OrderProductInStockEligibility $constraint */
        Assert::isInstanceOf($constraint, OrderProductInStockEligibility::class);
```

File: features/inventory/checkout_inventory/validating_product_quantity_on_checkout_complete.feature [hunk 15]
```diff
@@ -16,23 +16,23 @@ Feature: Being unable to buy products that are out of stock
         And the store allows paying offline
         And I am a logged in customer
 
-    @ui
+    @ui @api
     Scenario: Placing an order with products that have sufficient quantity
         Given I have added 3 products "Iron Maiden T-Shirt" to the cart
-        And I have proceeded selecting "Offline" payment method
+        And I proceed through checkout process
```
Comment (GSadee) [inline features/inventory/checkout_inventory/validating_product_quantity_on_checkout_complete.feature:22]:
```suggestion
        When I proceed through checkout process
```
but why did you change that step?

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 16]
```diff
@@ -72,6 +73,9 @@ final class CheckoutContext implements Context
     /** @var SharedStorageInterface */
     private $sharedStorage;
 
+    /** @var DefaultProductVariantResolver */
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:76]:
```suggestion
    /** @var ProductVariantResolverInterface */
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibility.php [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/** @experimental */
+final class OrderProductInStockEligibility extends Constraint
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductInStockEligibility.php:19]:
```suggestion
final class OrderProductAvailability extends Constraint
```
or
```suggestion
final class OrderItemAvailability extends Constraint
```

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    function it_adds_violation_if_product_variant_does_not_have_sufficient_stock(
+        ExecutionContextInterface $executionContext,
+        OrderRepositoryInterface $orderRepository,
+        OrderInterface $order,
+        OrderItemInterface $orderItem,
+        ProductVariantInterface $productVariant,
+        AvailabilityCheckerInterface $availabilityChecker,
+        Collection $orderItems
+    ): void {
+        $this->initialize($executionContext);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php:62]: This line can be moved to `let()` method

File: src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [
+                new CompleteOrder(),
+                new class() extends Constraint {}
+            ])
+        ;
+    }
+
+    function it_adds_violation_if_product_variant_does_not_have_sufficient_stock(
+        ExecutionContextInterface $executionContext,
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/OrderProductInStockEligibilityValidatorSpec.php:54]: I would be for keeping the same order of arguments as in `let()` method

File: features/inventory/checkout_inventory/validating_product_quantity_on_checkout_complete.feature (hunk 15 above)
Comment (AdamKasp) [reply features/inventory/checkout_inventory/validating_product_quantity_on_checkout_complete.feature:22]: I had a problem with addresses, in behats they were empty, and I get SM error, with this step, the scenario is still valid and checkout works as I expect.

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Adam! :tada:
